src/signatures.md

```diff
@@ -264,3 +264,72 @@ it should take `self` by value. Examples:
 
 * Closing a file and returning a result code.
 * A builder-pattern object which spits out the thing it was building. ([Example](https://docs.rs/bindgen/0.59.0/bindgen/struct.Builder.html#method.generate)).
+
+## How do I take a thing, and a reference to something within that thing?
+
+For example, suppose you want to give all of your dogs to your friend, yet also
+tell your friend which one of the dogs is the Best Boy or Girl.
+
+A common use case is if you find yourself destructuring a complex data
+structure to work out what to do with it, then you want to pass the whole thing
+into another function but you don't want to have to repeat all that
+destructuring work.
+
+Generally this is an indication that your types are not split down in the correct
+way:
+
+> This is a decomposition problem. Once you’ve found the correct decomposition, everything
+> else just works. The code almost writes itself. - AF.
+
+If you must do this though, perhaps you can pass the thing, and a closure
+to extract part of the thing.
+
+## When should I return `impl Trait`?
+
+Your main consideration should be API stability. If your caller doesn't
+_need_ to know the concrete implementation type, then don't tell it. That
+gives you flexibility to change your implementation in future without breaking
+compatibility.
+
+Note [Hyrum's Law](https://www.hyrumslaw.com/)!
```

Comment:
Tell people what about it they should take note of. For example, give people some advice on not leaking irrelevant aspects of the implementation type. This can mean

* minimality (e.g. return or take `Read`, not `File`, where possible — this also gives us greater generality), or
* time- or space-independence (e.g. if it's O(n), either guarantee it or hide it or disclaim it), or
* freedom from side-effects (e.g. if `Time::now` touches a file, someone's going to start depending on that as a spacebar heater (https://xkcd.com/1172/)).

Reply:
I agree with you, but would you mind writing this up as a separate pull request?

And - where possible - I'm trying to be Rust-specific, or at least alert people to Rusty things that they might not have thought about back in C++-land.